Aspect and Type api implementation

dhrn opened the pull request:
https://alfresco.atlassian.net/browse/REPO-5440
https://alfresco.atlassian.net/browse/REPO-5552

depends https://github.com/Alfresco/rest-api-explorer/pull/119

skopf reviewed (changes requested):
LGTM, just one small naming issue:
The abstract base class for the "Aspect" in the rest api package is "AbstractModel". What was the idea behind this base class? Where else are you planning to re-use this base class?
The common parent of aspects and types would be a "class" in alfresco speak. So "AbstractClass" would be a better fit.
Also I am trying to understand how you decided what parts should go into the "AbstractModel" and what parts should go into the "Aspect"

gavincornwell commented on remote-api/src/main/java/org/alfresco/rest/api/Aspects.java:
```diff
@@ -0,0 +1,37 @@
+ * #%L
+ * Alfresco Remote API
+ * %%
+ * Copyright (C) 2005 - 2016 Alfresco Software Limited
```
Please update the copyright year from 2016 to 2021, same comment for all new files below.

gavincornwell commented on remote-api/src/main/java/org/alfresco/rest/api/impl/AspectsImpl.java:
```diff
@@ -0,0 +1,321 @@
+        return null;
+    }
+
+    private Collection<QName> getModelAspects(String modelId) {
```
Please be consistent with the coding standards of this project.

gavincornwell commented on remote-api/src/main/java/org/alfresco/rest/api/impl/AspectsImpl.java:
```diff
@@ -0,0 +1,321 @@
+    private Collection<QName> getModelAspects(String modelId) {
+        ModelDefinition modelDefinition =  null;
+
+        if(modelId == null)
```
Please add a space between the if and opening bracket to be consistent with coding standards for this project.

gavincornwell reviewed (changes requested):
Nothing to add to what Stefan said except for a couple of minor coding standards items.

dhrn commented:
Similar pattern was followed for the CMM end point. i just followed it. 

Now we have `AbstractClass` that have common properties for aspect and type. 
now it is look all the properties are same for aspect and types. 

let me know the comments
Thanks

dhrn commented:
Here i have adding the slack discussion for other reviewers

current representation  contains `id` `title` `descrption`  `parentId`. previously the convention was `name` `parentName` . so  i can't use the old model. so I have created the new abstract class.

As per the above  i have renamed the AbstractModel to AbstractClass. now after this Here we have two models in the same package

-  `AbstractClass.java`
-  `AbstractClassModel.java` -> used in CMM API


This is little confusing to user. so we renamed `AbstractClassModel` to `AbstractCustomClass` to keep the consistency

after that, we found the `ClassDefinition` was used in Node. so it has properties that can be shared across the 
Aspect and Type . so we have extended the `AbstractClass` from `ClassDefinition` .

here is final implementation of the [AbstractClass](https://github.com/Alfresco/alfresco-community-repo/blob/dev-dhrn-REPO-5440/remote-api/src/main/java/org/alfresco/rest/api/model/AbstractClass.java)

gavincornwell reviewed (approved):
LGTM

gavincornwell commented on remote-api/src/main/java/org/alfresco/rest/api/impl/AbstractClassImpl.java:
```diff
@@ -0,0 +1,194 @@
+        }
+    }
+
+    public boolean filterByNamespace(ModelApiFilter query, QName type)
```
"type" is a mis-leading name for this parameter, "qname" would be better.

skopf reviewed (changes requested):
As Jan and Gav already mentioned on the api-explorer commit, we need to be really carful with naming. Since we are using the term "prefix" for this registered namespace prefix (e.g. "cm"), we can easily confuse readers if we use this term "prefix" as well for the prefix of the namespace URI.

gavincornwell commented on remote-api/src/main/java/org/alfresco/rest/api/impl/AbstractClassImpl.java:
```diff
@@ -0,0 +1,194 @@
+public class AbstractClassImpl<T extends AbstractClass> {
+    static String PARAM_MODEL_IDS = "modelIds";
+    static String PARAM_PARENT_IDS = "parentIds";
+    static String PARAM_URI_PREFIX = "uriPrefix";
```
As mentioned in the review of the open api spec, this should really be "namespaceUri) as that's what we're checking see lines 82 and 86.

gavincornwell commented on remote-api/src/main/java/org/alfresco/rest/api/impl/CustomModelsImpl.java:
```diff
@@ -1,28 +1,28 @@
+ * #%L
+ * Alfresco Remote API
+ * %%
+ * Copyright (C) 2005 - 2016 Alfresco Software Limited
```
Remember your copyright dates ;-)

gavincornwell reviewed (approved):
LGTM
